tools/java/org/unicode/cldr/api/PathMatcher.java

```diff
@@ -0,0 +1,276 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * An immutable matcher for {@link CldrPath} instances. A path matcher specification looks like
+ * {@code "//ldml/foo/*[@x="z"]/bar[@y=*]"}, where element names and attribute values can be
+ * wildcards. No capturing of wildcard values is done, and currently wildcards are limited to the
+ * entire field (element name, or attribute value).
+ *
+ * Examples of use:
+ *
+ * <h2>Matching an exact path</h2>
+ * <pre>{@code
+ * private static final PathMatcher ALIAS_TO_XX =
+ *     PathMatcher.of("//supplementalData/likelySubtags/likelySubtag[@from=*][@to="xx"]");
+ *
+ * if (ALIAS_TO_XX.matches(value.getPath())) {
+ *     // Value is a subtag alias to locale ID "xx".
+ * }
+ * }</pre>
+ *
+ * <h2>Matching path prefixes and hierarchical matching</h2>
+ * <pre>{@code
+ * private static final PathMatcher KEY = PathMatcher.of("//ldmlBCP47/keyword/key[@name=*]");
+ * private static final PathMatcher TYPE = key.extendBy("type[@name=*]");
+ *
+ * if (KEY.matchesPrefixOf(path)) {
+ *     if (TYPE.locallyMatches(path)) {
+ *         // Value is a BCP47 type.
+ *     }
+ * }
+ * }</pre>
+ */
+public final class PathMatcher {
+    private static final Pattern ROOT_PATH_SPEC = Pattern.compile("//(\\w+)(/|$)");
+
+    /**
+     * Parses the full or partial (prefix) path specification into a matcher. Path specifications
+     * given to this method must start with {@code "//xxx/..."} where {@code xxx} is a known CLDR
+     * XML path prefix (such as {@code "ldml").
+     */
+    public static PathMatcher of(String pathSpec) {
+        Matcher m = ROOT_PATH_SPEC.matcher(pathSpec);
+        checkArgument(m.lookingAt(), "invalid path specification: %s", pathSpec);
+        // This throws IllegalArgumentException if the type isn't valid.
+        CldrDataType.forXmlName(m.group(1));
+        return new PathMatcher(null, parse(pathSpec.substring(2)));
+    }
+
+    private final Optional<PathMatcher> parent;
+    private final ImmutableList<Predicate<CldrPath>> elementMatchers;
+    private final int totalElementCount;
+
+    private PathMatcher(PathMatcher parent, List<Predicate<CldrPath>> elementMatchers) {
+        this.parent = Optional.ofNullable(parent);
+        this.elementMatchers = ImmutableList.copyOf(elementMatchers);
+        this.totalElementCount =
+            elementMatchers.size() + (parent != null ? parent.totalElementCount : 0);
+    }
+
+    /** Extends this matcher by the given partial path specification. */
+    public PathMatcher extendBy(String pathSpec) {
+        return new PathMatcher(this, parse(pathSpec));
+    }
+
+    /**
+     * Returns whether this matcher fully matches the given path, including taking into account
+     * any parent matchers from which this matcher might have been extended.
+     */
+    public boolean matches(CldrPath path) {
+        if (!locallyMatches(path)) {
+            return false;
```

Comment:
Intuitively, the matching logic here seems reversed.
Should it not return true if locallyMatches()?
And if not, then false if there is no parent?

Reply:
If there's no parent, this is the root matcher, so the chain of matchers have all been matched successfully and we're done.
I added a comment to that effect.